File: zincc/src/parse/Parser.zig
allocator: Allocator,
tokens: []const TokenKind,
src: []const u8,

panicking: bool = false,
cursor: usize = 0,

errors: std.ArrayListUnmanaged(ParseError) = .{},

//

const std = @import("std");
const Allocator = std.mem.Allocator;
const Tokens = std.MultiArrayList(Token).Slice;

const super = @import("!mod.zig");

const Token = super.Token;
const TokenKind = super.Token.Kind;

const Cst = super.Cst;
const NodeBuilder = Cst.NodeBuilder;

const Self = @This();

const PathStartSet: []const TokenKind = &.{ .ident, .punct_dblColon };
const IntegerLiteralSet: []const TokenKind = &.{ .int_dec, .int_hex, .int_bin, .int_oct };

pub fn parse(allocator: Allocator, tokens: []const TokenKind, src: []const u8) ParseResult {
    var parser = Self{
        .allocator = allocator,
        .tokens = tokens,
        .src = src,
    };
    var root = parser.parseToplevel();
    defer root.deinit(allocator);
    return .{
        .cst = Cst.buildCst(allocator, root) catch unreachable,
        .errors = parser.errors,
    };
}

pub const ParseResult = struct {
    cst: Cst,
    errors: std.ArrayListUnmanaged(ParseError),

    pub fn deinit(self: *@This(), allocator: Allocator) void {
        self.cst.deinit(allocator);

        self.errors.deinit(allocator);
    }
};

pub const ParseError = union(enum) {
    expected: Expected,

    pub const Expected = struct {
        what: What,
        at: usize,
        found: usize,
        in: Context,

        pub const What = union(enum) {
            item: Item,
            token: TokenKind,
            one_of: []const What,

            pub const Item = enum {
                decl,
                expr,
            };
        };

        pub const Context = enum {
            top_level,
            decl_func,
            stmt,
            stmt_let,
            stmt_return,
            expr_start,
            expr_paren,
            expr_block,
            expr_call,
            string,
            path,
        };
    };
};

//
// Token ops
//

fn atEnd(self: Self) bool {
    return self.at(.EOF);
}

fn peekN(self: Self, n: usize) TokenKind {
    if (self.cursor + n >= self.tokens.len)
        return .EOF;
    return self.tokens[self.cursor + n];
}

fn peek(self: Self) TokenKind {
    return self.peekN(0);
}

// fn atN(self: Self, n: usize, kind: TokenKind) bool {
//     return self.peekN(n) == kind;
// }

fn at(self: Self, kind: TokenKind) bool {
    return self.peek() == kind;
}

// fn atSetN(self: Self, n: usize, set: []const TokenKind) bool {
//     for (set) |kind| {
//         if (self.atN(n, kind))
//             return true;
//     }
//     return false;
// }

fn atSet(self: Self, set: []const TokenKind) bool {
    for (set) |kind| {
        if (self.at(kind))
            return true;
    }
    return false;
    // return self.atSetN(0);
}

fn match(self: *Self, kind: TokenKind, parent: *NodeBuilder) bool {
    if (self.at(kind)) {
        self.bump(parent);
        return true;
    }
    return false;
}

fn matchSet(self: *Self, set: []const TokenKind, parent: *NodeBuilder) bool {
    if (self.atSet(set)) {
        self.bump(parent);
        return true;
    }
    return false;
}

//
// Error ops
//

fn report(self: *Self, parent: *NodeBuilder, err: ParseError) void {
    if (!self.panicking) {
        self.panicking = true;

        self.errors.append(self.allocator, err) catch unreachable;
        return;
    }

    if (self.atSet(&[_]TokenKind{
        .brkt_brace_close,
        .kw_fn,
        .EOF,
    })) {
        self.bump(parent);
        self.panicking = false;
    } else {
        var node = NodeBuilder{ .kind = .skipped };
        self.bump(&node);
        self.appendNodeTo(node, parent);
    }
}

fn expect(
    self: *Self,
    what: TokenKind,
    in: ParseError.Expected.Context,
    parent: *NodeBuilder,
) void {
    if (!self.match(what, parent)) {
        self.report(parent, .{
            .expected = .{
                .what = .{ .token = what },
                .at = self.cursor - 1,
                .found = self.cursor,
                .in = in,
            },
        });
    }
}

//
// Node ops
//

fn bump(self: *Self, parent: *NodeBuilder) void {
    self.appendTokenTo(parent);
    self.cursor += 1;
}

fn appendNodeTo(self: *Self, node: NodeBuilder, to: *NodeBuilder) void {
    to.appendNode(self.allocator, node) catch unreachable;
}

fn appendTokenTo(self: *Self, to: *NodeBuilder) void {
    to.appendToken(self.allocator) catch unreachable;
}

//
// Parse
//

/// Parse the top level of a module/file
fn parseToplevel(self: *Self) NodeBuilder {
    var root = NodeBuilder{ .kind = .root };

    while (!self.atEnd()) {
        self.parseDecl(.top_level, &root);
    }

    // // @FIXME: Is this needed?
    // while (self.tokens[self.cursor] != .EOF) {
    //     self.appendTokenTo(&root);
    //     self.cursor += 1;
    // }
    self.appendTokenTo(&root);

    return root;
}

// @FIXME: 'foo::::' case crashing
fn parsePath(self: *Self) NodeBuilder {
    if (!self.panicking)
        std.debug.assert(self.atSet(PathStartSet));

    var path = NodeBuilder{ .kind = .path };

    _ = self.match(.punct_dblColon, &path);
    self.expect(.ident, .path, &path);

    while (self.at(.punct_dblColon)) {
        self.expect(.punct_dblColon, .path, &path);
        self.expect(.ident, .path, &path);
    }

    return path;
}

fn parseString(self: *Self) NodeBuilder {
    std.debug.assert(self.at(.string_open));

    var str = NodeBuilder{ .kind = .string };

    self.bump(&str);
    while (self.matchSet(&.{
        .string_literal,
        .esc_char,
        .esc_asciicode,
        .esc_unicode,
    }, &str)) {}
    // Any unclosed strings should be caught by the lexer,
    // there's not much a compiler can do to try recover from a malformed string
    self.expect(.string_close, .string, &str);

    return str;
}

fn parseLiteralInt(self: *Self) NodeBuilder {
    std.debug.assert(self.atSet(IntegerLiteralSet));
    var lit = NodeBuilder{ .kind = .literal_int };
    self.bump(&lit);
    return lit;
}

fn parseLiteralFloat(self: *Self) NodeBuilder {
    std.debug.assert(self.at(.float));
    var lit = NodeBuilder{ .kind = .literal_float };
    self.bump(&lit);
    return lit;
}

//
// Parse Decl
//

fn parseDecl(self: *Self, in: ParseError.Expected.Context, parent: *NodeBuilder) void {
    if (!self.tryParseDecl(parent)) {
        self.report(parent, .{
            .expected = .{
                .what = .{ .item = .decl },
                .at = self.cursor,
                .found = self.cursor,
                .in = in,
            },
        });
    }
}

fn tryParseDecl(self: *Self, parent: *NodeBuilder) bool {
    if (self.at(.ident)) {
        const p_next = self.peekN(1);
        var node = NodeBuilder{ .kind = undefined };
        switch (p_next) {
            .kw_fn => {
                node.kind = .decl_func;
                self.bump(&node); // ident
                self.parseDeclFunction(&node);
            },
            .kw_const => {
                node.kind = .decl_const;
                self.bump(&node); // ident
                self.bump(&node); // kw_const
                self.parseStmtExpr(&node);
            },
            else => return false,
        }
        self.appendNodeTo(node, parent);
        return true;
    }
    return false;
}

fn parseDeclFunction(self: *Self, func: *NodeBuilder) void {
    std.debug.assert(self.at(.kw_fn));

    {
        var sig = NodeBuilder{ .kind = .func_signature };
        defer self.appendNodeTo(sig, func);

        self.bump(&sig); // kw_fn

        if (self.match(.brkt_paren_open, &sig)) {
            if (!self.at(.brkt_paren_close)) {
                while (true) {
                    if (self.at(.brkt_paren_close))
                        break;

                    var arg = NodeBuilder{ .kind = .func_signature_arg };
                    defer self.appendNodeTo(arg, &sig);

                    self.expect(.ident, .decl_func, &arg);
                    self.parseType(&arg);

                    if (!self.match(.punct_comma, &sig))
                        break;
                }
            }

            self.expect(.brkt_paren_close, .decl_func, &sig);
        }

        if (!self.atSet(&.{ .punct_fat_arrow, .brkt_brace_open })) {
            var ret = NodeBuilder{ .kind = .func_signature_ret };
            defer self.appendNodeTo(ret, &sig);
            self.parseType(&ret);
        }
    }

    var body = NodeBuilder{ .kind = .decl_func_body };
    defer self.appendNodeTo(body, func);

    if (self.match(.punct_fat_arrow, &body)) {
        self.parseStmtExpr(&body);
    } else {
        self.appendNodeTo(self.parseExprBlock(), &body);
    }
}

//
// Parse type
//

fn parseType(self: *Self, parent: *NodeBuilder) void {
    self.appendNodeTo(self.parsePath(), parent);
}

//
// Parse stmt
//

fn parseStmt(self: *Self, parent: *NodeBuilder) void {
    switch (self.peek()) {
        .kw_let => self.parseStmtLet(parent),
        .kw_return => {
            var ret = NodeBuilder{ .kind = .stmt_return };
            defer self.appendNodeTo(ret, parent);
            self.bump(&ret);
            if (!self.at(.punct_semiColon)) {
                var expr = NodeBuilder{ .kind = .stmt_return_expr };
                defer self.appendNodeTo(expr, &ret);
                self.parseExpr(&expr);
            }
            self.expect(.punct_semiColon, .stmt_return, &ret);
        },
        else => self.parseStmtExpr(parent),
    }
}

fn parseStmtExpr(self: *Self, parent: *NodeBuilder) void {
    self.parseExpr(parent);

    if (self.tokens[self.cursor - 1] != .brkt_brace_close) {
        self.expect(.punct_semiColon, .stmt, parent);
    }
}

fn parseStmtLet(self: *Self, parent: *NodeBuilder) void {
    std.debug.assert(self.at(.kw_let));

    var let = NodeBuilder{ .kind = .stmt_let };
    defer self.appendNodeTo(let, parent);
    self.bump(&let); // kw

    self.expect(.ident, .stmt_let, &let); // ident

    if (self.match(.punct_colon, &let)) {
        var ty = NodeBuilder{ .kind = .stmt_let_ty };
        defer self.appendNodeTo(ty, &let);
        self.parseType(&ty);
    }

    self.expect(.punct_eq, .stmt_let, &let); // '='

    if (!self.at(.punct_semiColon)) {
        var expr = NodeBuilder{ .kind = .stmt_let_expr };
        defer self.appendNodeTo(expr, &let);

        self.parseExpr(&expr);
    }

    self.expect(.punct_semiColon, .stmt_let, &let);
}

//
// Parse expr
//

fn parseExpr(self: *Self, parent: *NodeBuilder) void {
    self.parseExprPrecedence(.none, parent);
}

const Precedence = enum(u8) {
    none = 0,
    assignment,
    // logical_or,
    // logical_and,
    sum,
    product,
    // prefix,
    call,

    pub fn get(kind: TokenKind) Precedence {
        return switch (kind) {
            .punct_eq => .assignment,
            // .Kw_or => .logical_or,
            // .Kw_and => .logical_and,
            .punct_plus, .punct_minus => .sum,
            .punct_star, .punct_slash => .product,
            // .Punct_Exclamation, .Kw_not => .prefix,
            .brkt_paren_open => .call,
            // .ident, .punct_dblColon => .call, // infix call
            else => .none,
        };
    }
};

fn parseExprPrecedence(self: *Self, precedence: Precedence, parent: *NodeBuilder) void {
    var lhs = if (self.parseExprStart()) |n| n else {
        self.report(parent, .{
            .expected = .{
                .what = .{ .item = .expr },
                .at = self.cursor,
                .found = self.cursor,
                .in = .expr_start,
            },
        });
        return;
    };

    const initial_prec = @enumToInt(precedence);
    var p = self.peek();
    var infix_prec = @enumToInt(Precedence.get(self.peek()));

    while (initial_prec < infix_prec) {
        if (self.parseExprInfix(p, lhs, infix_prec)) |node|
            lhs = node
        else {
            // @FIXME: Is this an error condition?
            self.appendNodeTo(lhs, parent);
            return;
        }

        p = self.peek();
        infix_prec = @enumToInt(Precedence.get(p));
    }

    self.appendNodeTo(lhs, parent);
}

fn parseExprStart(self: *Self) ?NodeBuilder {
    return switch (self.peek()) {
        .ident, .punct_dblColon => self.parsePath(),
        .string_open => self.parseString(),

        .int_dec, .int_hex, .int_bin, .int_oct => self.parseLiteralInt(),
        .float => self.parseLiteralFloat(),

        .brkt_brace_open => self.parseExprBlock(),

        // .brkt_paren_open => blk: {
        //     var expr = NodeBuilder{ .kind = undefined };

        //     self.bump(&expr); // '('

        //     // unit
        //     if (self.match(.brkt_paren_close, &expr)) { // ')'
        //         expr.kind = .expr_unit;
        //         break :blk expr;
        //     }
        // },

        else => return null,
    };
}

fn parseExprBlock(self: *Self) NodeBuilder {
    std.debug.assert(self.at(.brkt_brace_open));

    var block = NodeBuilder{ .kind = .expr_block };

    self.bump(&block); // '{'

    while (!self.atSet(&.{ .brkt_brace_close, .EOF })) {
        self.parseStmt(&block);
    }

    self.expect(.brkt_brace_close, .expr_block, &block); // '}'

    return block;
}

fn parseExprInfix(self: *Self, p: TokenKind, lhs: NodeBuilder, prec: u8) ?NodeBuilder {
    return switch (p) {
        .punct_eq,
        .punct_plus,
        .punct_minus,
        .punct_star,
        .punct_slash,
        => blk: {
            var infix = NodeBuilder{ .kind = .expr_infix };

            { // lhs
                var left = NodeBuilder{ .kind = .expr_infix_lhs };
                defer self.appendNodeTo(left, &infix);
                self.appendNodeTo(lhs, &left); // lhs
            }

            { // op
                var op = NodeBuilder{ .kind = .expr_infix_op };
                defer self.appendNodeTo(op, &infix);

                // @TODO: Parse multiple operators
                self.bump(&op); // operator
            }

            { // rhs
                var right = NodeBuilder{ .kind = .expr_infix_rhs };
                defer self.appendNodeTo(right, &infix);
                self.parseExprPrecedence(@intToEnum(Precedence, prec), &right); // rhs
            }

            break :blk infix;
        },
        .brkt_paren_open => blk: {
            var call = NodeBuilder{ .kind = .expr_call };

            {
                var callee = NodeBuilder{ .kind = .expr_call_callee };
                defer self.appendNodeTo(callee, &call);
                self.appendNodeTo(lhs, &callee);
            }

            self.bump(&call); // '('

            if (self.match(.brkt_paren_close, &call)) // ')'
                break :blk call;

            {
                var args = NodeBuilder{ .kind = .expr_call_args };
                defer self.appendNodeTo(args, &call);

                self.parseExpr(&args);

                while (self.match(.punct_comma, &args)) {
                    if (self.at(.brkt_paren_close))
                        break;
                    self.parseExpr(&args);
                }
            }

            self.expect(.brkt_paren_close, .expr_call, &call); // ')'

            break :blk call;
        },
        else => return null,
    };
}

File: zincc/src/parse/parser.rs
use super::{
    cst::{self, NodeKind, NK},
    TokenKind, TK,
};

pub fn parse(tokens: &[TokenKind]) -> ParseResult {
    let mut parser = Parser {
        tokens,

        panicking: false,
        cursor: 0,

        errors: Vec::new(),

        node_map: Default::default(),
    };
    let root = parser.parse_top_level();
    let mut map = parser.node_map;
    let root = map.push(root.node);
    ParseResult {
        cst: cst::Cst {
            root: cst::NamedNodeId {
                raw: root,
                kind: NK::root,
            },
            map,
        },
        errors: parser.errors,
    }
}

struct Parser<'s> {
    tokens: &'s [TokenKind],

    panicking: bool,
    cursor: usize,

    errors: Vec<ParseError>,

    node_map: crate::util::index::IndexVec<cst::Node, cst::NodeId>,
}

#[derive(Debug)]
pub struct ParseResult {
    pub cst: cst::Cst,
    pub errors: Vec<ParseError>,
}

#[derive(Debug)]
pub enum ParseError {
    Expected(ParseErrorExpected),
}

#[derive(Debug)]
pub struct ParseErrorExpected {
    pub what: ParseErrorExpectedWhat,
    pub at: usize,
    pub found: usize,
    pub context: ParseContext,
}

#[derive(Debug)]
pub enum ParseErrorExpectedWhat {
    Item(ParseErrorItem),
    Token(TK),
    // OneOf(Vec<ParseErrorExpectedWhat>),
}

#[derive(Debug)]
pub enum ParseErrorItem {
    Decl,
    Expr,
    Ty,
}

#[derive(Debug, Clone, Copy)]
pub enum ParseContext {
    TopLevel,
    Block,
    DeclFunc,
    DeclConst,
    Stmt,
    // StmtLet,
    // StmtReturn,
    ExprStart,
    // ExprParen,
    ExprCall,
    String,
    Path,
    TyFunc,
    Ty,
    TySlice,
}

#[derive(Clone)]
struct PNode {
    kind: NodeKind,
    node: cst::Node,
}

/// Token ops
impl Parser<'_> {
    fn at_end(&self) -> bool {
        self.at(TK::Eof)
    }

    fn peek_n(&self, n: usize) -> TK {
        if self.cursor + n >= self.tokens.len() {
            TK::Eof
        } else {
            self.tokens[self.cursor + n]
        }
    }

    fn peek(&self) -> TK {
        self.peek_n(0)
    }

    fn at(&self, kind: TK) -> bool {
        self.peek() == kind
    }

    fn at_set(&self, set: &[TK]) -> bool {
        for kind in set {
            if self.at(*kind) {
                return true;
            }
        }
        false
    }

    fn eat(&mut self, kind: TK, parent: &mut PNode) -> bool {
        if self.at(kind) {
            self.bump(parent);
            true
        } else {
            false
        }
    }

    fn eat_set(&mut self, set: &[TK], parent: &mut PNode) -> bool {
        if self.at_set(set) {
            self.bump(parent);
            true
        } else {
            false
        }
    }
}

/// Node ops
impl Parser<'_> {
    fn bump(&mut self, parent: &mut PNode) {
        parent
            .node
            .elements
            .push(cst::Element::Token(cst::TokenIndex::new(self.cursor)));
        self.cursor += 1;
    }

    fn node(&self, kind: NK) -> PNode {
        PNode {
            kind,
            node: cst::Node::new(),
        }
    }

    fn append_node(&mut self, what: PNode, to: &mut PNode) {
        let id_raw = self.node_map.push(what.node);
        let id = cst::NamedNodeId {
            raw: id_raw,
            kind: what.kind,
        };
        to.node.elements.push(cst::Element::Node(id));
    }

    fn node_bump(&mut self, nk: NK) -> PNode {
        let mut n = self.node(nk);
        self.bump(&mut n);
        n
    }
}

/// Error ops
impl Parser<'_> {
    fn report(&mut self, parent: &mut PNode, err: ParseError) {
        if !self.panicking {
            self.panicking = true;

            self.errors.push(err);
            return;
        }

        if self.at_set(&[TK::brkt_brace_close, TK::Eof, TK::kw_let]) {
            self.bump(parent);
            self.panicking = false;
        } else {
            let mut node = self.node(NK::skipped);
            self.bump(&mut node);
            self.append_node(node, parent);
        }
    }

    fn expect(&mut self, what: TK, context: ParseContext, parent: &mut PNode) {
        if !self.eat(what, parent) {
            self.report(
                parent,
                ParseError::Expected(ParseErrorExpected {
                    what: ParseErrorExpectedWhat::Token(what),
                    at: self.cursor - 1,
                    found: self.cursor,
                    context,
                }),
            )
        }
    }
}

/// Parse
impl Parser<'_> {
    fn parse_top_level(&mut self) -> PNode {
        let mut root = self.node(NK::root);

        while !self.at_end() {
            self.parse_decl(ParseContext::TopLevel, &mut root);
        }

        self.bump(&mut root);

        root
    }

    fn parse_path(&mut self) -> PNode {
        let mut path = self.node(NK::path);

        self.eat(TK::punct_dblColon, &mut path);
        self.expect(TK::ident, ParseContext::Path, &mut path);

        while self.at(TK::punct_dblColon) {
            self.expect(TK::punct_dblColon, ParseContext::Path, &mut path);
            self.expect(TK::ident, ParseContext::Path, &mut path);
        }

        path
    }

    fn parse_string(&mut self) -> PNode {
        assert!(self.at(TK::string_open));

        let mut str = self.node(NK::string);

        self.bump(&mut str);
        while self.eat_set(
            &[
                TK::string_literal,
                TK::esc_asciicode,
                TK::esc_unicode,
                TK::esc_asciicode,
                TK::esc_unicode,
                TK::esc_char_newline,
                TK::esc_char_return,
                TK::esc_char_tab,
                TK::esc_char_backslash,
                TK::esc_char_doublequote,
                TK::esc_char_singlequote,
                TK::esc_char_other,
            ],
            &mut str,
        ) {}
        self.expect(TK::string_close, ParseContext::String, &mut str);

        str
    }

    fn parse_literal_int(&mut self) -> PNode {
        assert!(self.at_set(&[TK::int_dec, TK::int_bin, TK::int_hex, TK::int_oct]));
        let mut lit = self.node(NK::literal_int);
        self.bump(&mut lit);
        lit
    }

    fn parse_literal_float(&mut self) -> PNode {
        assert!(self.at(TK::float));
        let mut lit = self.node(NK::literal_float);
        self.bump(&mut lit);
        lit
    }

    fn parse_block(&mut self) -> PNode {
        let mut block = self.node(NK::block);

        self.expect(TK::brkt_brace_open, ParseContext::Block, &mut block);

        while !self.at_set(&[TK::brkt_brace_close, TK::Eof]) {
            self.parse_stmt(&mut block);
        }

        self.expect(TK::brkt_brace_close, ParseContext::Block, &mut block);

        block
    }

    /// 'let's and 'const's
    fn parse_binding(&mut self, binding: &mut PNode) {
        // ident
        self.expect(TK::ident, ParseContext::DeclConst, binding);

        // ( ':' ty )?
        if self.eat(TK::punct_colon, binding) {
            let mut ty = self.node(NK::binding_ty);
            self.parse_ty(&mut ty);
            self.append_node(ty, binding);
        }

        // '=' expr
        self.expect(TK::punct_eq, ParseContext::DeclConst, binding);
        self.parse_stmt_expr(binding);
    }
}

/// Parse decl
impl Parser<'_> {
    fn parse_decl(&mut self, context: ParseContext, parent: &mut PNode) {
        if let Some(decl) = self.try_parse_decl() {
            self.append_node(decl, parent);
        } else {
            self.report(
                parent,
                ParseError::Expected(ParseErrorExpected {
                    what: ParseErrorExpectedWhat::Item(ParseErrorItem::Decl),
                    at: self.cursor,
                    found: self.cursor,
                    context,
                }),
            )
        }
    }

    fn try_parse_decl(&mut self) -> Option<PNode> {
        let decl = match self.peek() {
            TK::kw_fn => {
                let mut func = self.node(NK::decl_func);
                self.bump(&mut func); // kw_fn
                self.bump(&mut func); // ident

                let mut proto = self.node(NK::func_proto);
                self.parse_func_params_and_ret(ParseContext::DeclFunc, &mut proto);
                self.append_node(proto, &mut func);

                let mut body = self.node(NK::decl_func_body);

                if self.eat(TK::punct_fat_arrow, &mut body) {
                    self.parse_stmt_expr(&mut body);
                } else {
                    let block = self.parse_block();
                    self.append_node(block, &mut body);
                }

                self.append_node(body, &mut func);
                func
            }
            TK::kw_const => {
                let mut konst = self.node(NK::decl_const);

                self.bump(&mut konst); // 'const'
                self.parse_binding(&mut konst);

                konst
            }
            _ => return None,
        };
        Some(decl)
    }
}

/// Parse ty
impl Parser<'_> {
    fn parse_ty(&mut self, parent: &mut PNode) {
        match self.peek() {
            TK::ident | TK::punct_dblColon => {
                let path = self.parse_path();
                self.append_node(path, parent);
            }
            TK::brkt_square_open => {
                let mut slice_ty = self.node(NK::ty_slice);
                self.bump(&mut slice_ty); // '['
                self.expect(TK::brkt_square_close, ParseContext::TySlice, &mut slice_ty); // ']'
                self.parse_ty(&mut slice_ty); // ty
                self.append_node(slice_ty, parent);
            }
            TK::punct_question => {
                let mut nullable_ty = self.node(NK::ty_nullable);
                self.bump(&mut nullable_ty); // '?'
                self.parse_ty(&mut nullable_ty); // ty
                self.append_node(nullable_ty, parent);
            }
            TK::kw_fn => self.parse_ty_func(parent),
            _ => self.report(
                parent,
                ParseError::Expected(ParseErrorExpected {
                    what: ParseErrorExpectedWhat::Item(ParseErrorItem::Ty),
                    at: self.cursor,
                    found: self.cursor,
                    context: ParseContext::Ty,
                }),
            ),
        }
    }

    fn parse_func_params_and_ret(&mut self, ctx: ParseContext, parent: &mut PNode) {
        self.expect(TK::brkt_paren_open, ctx, parent);
        while !self.at(TK::brkt_paren_close) {
            let mut param = self.node(NK::func_proto_param);

            if self.at(TK::ident) && self.peek_n(1) == TK::punct_colon {
                self.expect(TK::ident, ctx, &mut param);
                self.expect(TK::punct_colon, ctx, &mut param);
            }

            self.parse_ty(&mut param);
            self.append_node(param, parent);

            if !self.eat(TK::punct_comma, parent) {
                break;
            }
        }
        self.expect(TK::brkt_paren_close, ctx, parent);

        if self.eat(TK::punct_colon, parent) {
            let mut ret = self.node(NK::func_proto_ret);
            self.parse_ty(&mut ret);
            self.append_node(ret, parent);
        }
    }

    fn parse_ty_func(&mut self, parent: &mut PNode) {
        assert!(self.at(TK::kw_fn));

        let mut proto = self.node(NK::func_proto);
        self.bump(&mut proto); // 'fn'

        self.parse_func_params_and_ret(ParseContext::TyFunc, &mut proto);

        self.append_node(proto, parent);
    }
}

/// Parse stmt
impl Parser<'_> {
    fn parse_stmt(&mut self, parent: &mut PNode) {
        match self.peek() {
            TK::kw_let => {
                let mut binding = self.node(NK::stmt_let);

                self.bump(&mut binding); // 'let'

                self.parse_binding(&mut binding);

                self.append_node(binding, parent);
            }
            _ => {
                if let Some(decl) = self.try_parse_decl() {
                    let mut stmt = self.node(NK::stmt_decl);
                    self.append_node(decl, &mut stmt);
                    self.append_node(stmt, parent);
                } else {
                    let mut stmt = self.node(NK::stmt_expr);
                    self.parse_stmt_expr(&mut stmt);
                    self.append_node(stmt, parent);
                }
            }
        }
    }

    /// Parse an expression in the context of a statement.
    /// ie. with a semicolon if it does not end with a '}'
    fn parse_stmt_expr(&mut self, parent: &mut PNode) {
        self.parse_expr(parent);

        if self.tokens[self.cursor - 1] != TK::brkt_brace_close {
            self.expect(TK::punct_semiColon, ParseContext::Stmt, parent);
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
#[repr(u8)]
enum Precedence {
    None = 0,
    Assignment,
    LogicalOr,
    LogicalAnd,
    Sum,
    Product,
    // Suffix, // ? ! .
    Call,
}

impl Precedence {
    fn of(kind: TK) -> Self {
        match kind {
            TK::punct_eq => Self::Assignment,
            TK::kw_or => Self::LogicalOr,
            TK::kw_and => Self::LogicalAnd,
            TK::punct_plus | TK::punct_minus => Self::Sum,
            TK::punct_star | TK::punct_slash => Self::Product,
            TK::brkt_paren_open => Self::Call,
            _ => Self::None,
        }
    }
}

/// Parse expr
impl Parser<'_> {
    fn parse_expr(&mut self, parent: &mut PNode) {
        self.parse_expr_precedence(Precedence::None, parent);
    }

    fn try_parse_expr(&mut self) -> Option<PNode> {
        self.try_parse_expr_precedence(Precedence::None)
    }

    fn parse_expr_precedence(&mut self, prec: Precedence, parent: &mut PNode) {
        if let Some(expr) = self.try_parse_expr_precedence(prec) {
            self.append_node(expr, parent);
        } else {
            self.report(
                parent,
                ParseError::Expected(ParseErrorExpected {
                    what: ParseErrorExpectedWhat::Item(ParseErrorItem::Expr),
                    at: self.cursor,
                    found: self.cursor,
                    context: ParseContext::ExprStart,
                }),
            );
        }
    }

    fn try_parse_expr_precedence(&mut self, prec: Precedence) -> Option<PNode> {
        let mut lhs = self.try_parse_expr_start()?;

        let init_prec = prec;
        let mut p = self.peek();
        let mut infix_prec = Precedence::of(p);

        while init_prec < infix_prec {
            if let Some(e) = self.try_parse_expr_infix(p, lhs.clone(), infix_prec) {
                lhs = e;
            } else {
                // @FIXME:...
                //  Is this an error condition?
                //  Is it even reachable?
                // return Some(lhs);
                unreachable!();
            }

            p = self.peek();
            infix_prec = Precedence::of(p);
        }

        Some(lhs)
    }

    fn try_parse_expr_start(&mut self) -> Option<PNode> {
        let expr = match self.peek() {
            TK::ident => self.parse_path(),
            TK::string_open => self.parse_string(),

            TK::int_dec | TK::int_hex | TK::int_bin | TK::int_oct => self.parse_literal_int(),
            TK::float => self.parse_literal_float(),

            TK::kw_false => self.node_bump(NK::expr_false),
            TK::kw_true => self.node_bump(NK::expr_true),

            TK::brkt_brace_open => self.parse_block(),

            TK::brkt_paren_open => todo!("parse paren exprs"),

            TK::kw_return => {
                let mut ret = self.node(NK::expr_return);
                self.bump(&mut ret); // 'return'
                if let Some(expr) = self.try_parse_expr() {
                    self.append_node(expr, &mut ret);
                }
                ret
            }

            _ => return None,
        };
        Some(expr)
    }

    fn try_parse_expr_infix(&mut self, p: TK, lhs: PNode, prec: Precedence) -> Option<PNode> {
        let node = match p {
            TK::punct_eq | TK::punct_plus | TK::punct_minus | TK::punct_star | TK::punct_slash => {
                let mut infix = self.node(NK::expr_infix);

                // lhs
                self.append_node(lhs, &mut infix);

                // op
                {
                    let mut op = self.node(NK::expr_infix_op);
                    // @TODO: parse multiple operators
                    self.bump(&mut op);
                    self.append_node(op, &mut infix);
                }

                // rhs
                self.parse_expr_precedence(prec, &mut infix);

                infix
            }
            TK::brkt_paren_open => {
                let mut call = self.node(NK::expr_call);

                // callee
                self.append_node(lhs, &mut call);

                self.bump(&mut call); // '('

                // args
                if !self.at(TK::brkt_paren_close) {
                    self.parse_expr(&mut call);
                    while self.eat(TK::punct_comma, &mut call)
                        && !self.at(TokenKind::brkt_paren_close)
                    {
                        self.parse_expr(&mut call);
                    }
                }

                self.expect(TK::brkt_paren_close, ParseContext::ExprCall, &mut call); // ')'

                call
            }
            _ => return None,
        };
        Some(node)
    }
}
